### PR TITLE
Fix bugs identified in integration with go-data-transfer

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Integrate latest go-graphsync to go-data-transfer without issues arising. When first run, various data transfer integration tests involving network errors crashed.

# Implementation

Issue 1: Not cleaning up traversed links after network error causes response generation to fail

On the responding side, as we perform a selector traversal, we keep track of links traversed in generating a response, so as to avoiding sending blocks twice (if a single response traverses the same block twice, or multiple simultaneous responses traverse the same block, we should only send it once). However, when a network error occurred, we are not clearing the list of links traversed, as we would for any other error. Prior to the refactor, this didn't matter, because the link tracker was deleted entirely when peers disconnected. However, with our reduced-in-scope ResponseAssembler, we do not wipe out the link tracker when peers disconnect. As such, we need to actually cleanup requests that terminate on a network error. We renamed the `FinishWithCancel` method to `ClearRequest`. This method was previously called when a request is cancelled from the requestor side, but had no effect on the actual response other than to clear the links tracked cache. We now call the newly renamed `ClearRequest` method when a network error is detected causing a request to terminate.

Issue 2: Asynchronous block memory release can cause an overflow error to allocator if all peer memory is released and then individual block memory happens afterward, causing a double release.

This occurs in the MessageQueue when a peer disconnects. When a disconnect occurs, all outstanding messages in the queue dispatch an error notification for the message. Then all memory held by the peer is released in the allocator. Then seperately, an allocator subscriber receives the error notification and releases memory for the individual message. This may happen AFTER all memory for the peer is released. This can in some cases case memory to be freed with an unsigned int total of zero, causing and overflow which in turn puts the allocated memory artificially high.

The immediate fix is to avoid an unsigned int overflow in the allocator by just setting the total to zero when this happens, which I have done.

However, I think it's silly and potentially bad design (my own) to free message memory in a subscriber that runs asynchronously. As a result, I've refactored the message queue so that this happens synchronously, as soon as message processing is finished with a sent or an error.

#### other changes

In the process of fixing the allocator, I made two other changes.
1. I refactored the tests so they are much easier to understand -- now, everything is executed step wise - we run an allocation or deallocation, then we check to see everyone has the right memory we expect them to have, then we verify which requests are still pending due to total memory or per peer maxout.
2. I moved the allocator to a top level directory, given that it's now used solely in the MessageQueue.

### CODE REVIEW

File: allocator/allocator_test.go [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+					expectedPending: []pendingResult{{peers[1], 100}, {peers[2], 100}, {peers[2], 200}, {peers[1], 200}, {peers[2], 100}, {peers[1], 300}},
+				},
+				{
+					op:              alloc{peers[0], 0, true},
+					totals:          map[peer.ID]uint64{peers[0]: 0, peers[1]: 800, peers[2]: 900},
+					expectedPending: []pendingResult{{peers[1], 300}},
+				},
+			},
+		},
+		"release with nothing allocated": {
```
Comment (acruikshank) [inline allocator/allocator_test.go:287]: This seems misnamed. "release more than currently allocated"?

File: messagequeue/messagequeue.go [hunk 2]
```diff
@@ -301,22 +296,26 @@ func openSender(ctx context.Context, network MessageNetwork, p peer.ID) (gsnet.M
 	return nsender, nil
 }
 
-type allocatorSubscriber struct {
-	allocator Allocator
-	p         peer.ID
+type messagePublisher struct {
```
Comment (acruikshank) [inline messagequeue/messagequeue.go:299]: great refactor.

**Review by acruikshank — APPROVED** [review]
The changes look good. I made a few suggestions for clarity.

File: allocator/allocator_test.go [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+					totals:          map[peer.ID]uint64{peers[0]: 600},
+					expectedPending: []pendingResult{},
+				},
+				{
+					op:              alloc{peers[0], 300, false},
+					totals:          map[peer.ID]uint64{peers[0]: 900},
+					expectedPending: []pendingResult{},
+				},
+				{
+					op:              alloc{peers[0], 300, false},
```
[the remark this answers is not in the document]
Comment (hannahhoward) [reply allocator/allocator_test.go:41]: ??? we're allocating 1200 worth here? what we verify is after 3 adds of 300, the fourth add of 300 goes to pending (maxing at 900).

File: allocator/allocator_test.go [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+					totals:          map[peer.ID]uint64{peers[0]: 900, peers[1]: 900, peers[2]: 800},
+					expectedPending: []pendingResult{{peers[1], 400}, {peers[2], 300}},
+				},
+				{
+					op:              alloc{peers[2], 100, false},
+					totals:          map[peer.ID]uint64{peers[0]: 900, peers[1]: 900, peers[2]: 800},
+					expectedPending: []pendingResult{{peers[1], 400}, {peers[2], 300}, {peers[2], 100}},
+				},
+				{
+					op:              alloc{peers[1], 200, true},
```
[the remark this answers is not in the document]
Comment (hannahhoward) [reply allocator/allocator_test.go:205]: yes

**Pull request merged**